master/server.go

```diff
@@ -264,15 +242,98 @@ func (s *Server) startGrpcSrv() (err error) {
 	if err != nil {
 		return
 	}
-
 	log.L().Logger.Info("start etcd successfully")
 
 	// start grpc server
-
 	s.etcdClient, err = etcdutil.CreateClient([]string{withHost(s.cfg.MasterAddr)}, nil)
 	return
 }
 
+// name is a shortcut to etcd name
+func (s *Server) name() string {
+	return s.cfg.Etcd.Name
+}
+
+func (s *Server) reset(ctx context.Context) error {
+	sess, err := concurrency.NewSession(
+		s.etcdClient, concurrency.WithTTL(int(s.cfg.KeepAliveTTL.Seconds())))
+	if err != nil {
+		return errors.Wrap(errors.ErrMasterNewServer, err)
+	}
+	_, err = s.etcdClient.Put(ctx, adapter.MasterInfoKey.Encode(s.name()),
+		s.cfg.String(), clientv3.WithLease(sess.Lease()))
+	if err != nil {
+		return errors.Wrap(errors.ErrEtcdAPIError, err)
+	}
+
+	s.session = sess
+	s.election = concurrency.NewElection(sess, adapter.MasterCampaignKey.Path())
+	return nil
+}
+
+func (s *Server) runLeaderService(ctx context.Context) (err error) {
+	// rebuild states from existing meta if needed
+	err = s.resetExecutor(ctx)
+	if err != nil {
+		return
+	}
+
+	// start background managers
+	s.executorManager.Start(ctx)
+	err = s.jobManager.Start(ctx)
+	if err != nil {
+		return
+	}
+
+	s.leaderName.Store(s.name())
+	defer func() {
+		s.leaderName.Store("")
+		s.resign()
+	}()
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.session.Done():
+		return errors.ErrMasterSessionDone.GenWithStackByArgs()
+	}
+}
+
+func (s *Server) apiPreCheck() *pb.Error {
```

Comment:
if not the leader, the request will be  forwarded in this method ?

Reply:
The forward feature will be added later, it doesn't affect the main workflow. Another feature will be added is the gRPC client can find not leader and auto selection to real leader.